lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -13,548 +13,1055 @@
+      emitCircuitError()
+          << "contains a 'parent' with id '"
+          << a.first.cast<IntegerAttr>().getValue().getZExtValue()
+          << "', but does not contain a GrandCentral 'companion' "
+             "with the same id";
+      return signalPassFailure();
+    }
   }
 
-  // Populate interfaces.
-  for (auto &a : interfaceKeys) {
-    auto defName = a.first;
-    auto name = a.second;
+  LLVM_DEBUG({
+    // Print out the companion map, parent map, and all leaf values that
+    // were discovered.  Sort these by their keys before printing to make
+    // this easier to read.
+    SmallVector<IntegerAttr> ids;
+    auto sort = [&ids]() {
+      llvm::sort(ids, [](IntegerAttr a, IntegerAttr b) {
+        return a.getValue().getZExtValue() < b.getValue().getZExtValue();
+      });
+    };
+    for (auto tuple : companionIDMap)
+      ids.push_back(tuple.first.cast<IntegerAttr>());
+    sort();
+    llvm::dbgs() << "companionIDMap:\n";
+    for (auto id : ids) {
+      auto value = companionIDMap.lookup(id);
+      llvm::dbgs() << "  - " << id.getValue() << ": "
+                   << value.companion.getName() << " -> " << value.name << "\n";
+    }
+    llvm::dbgs() << "parentIDMap:\n";
+    for (auto id : ids) {
+      auto value = parentIDMap.lookup(id);
+      llvm::dbgs() << "  - " << id.getValue() << ": " << value.first.name()
+                   << ":" << value.second.getName() << "\n";
+    }
+    ids.clear();
+    for (auto tuple : leafMap)
+      ids.push_back(tuple.first.cast<IntegerAttr>());
+    sort();
+    llvm::dbgs() << "leafMap:\n";
+    for (auto id : ids) {
+      auto value = leafMap.lookup(id);
+      if (auto blockArg = value.dyn_cast<BlockArgument>()) {
+        FModuleOp module = cast<FModuleOp>(blockArg.getOwner()->getParentOp());
+        llvm::dbgs() << "  - " << id.getValue() << ": "
+                     << module.getName() + ">" +
+                            module.portNames()[blockArg.getArgNumber()]
+                                .cast<StringAttr>()
+                                .getValue()
+                     << "\n";
+      } else
+        llvm::dbgs() << "  - " << id.getValue() << ": "
+                     << value.getDefiningOp()
+                            ->getAttr("name")
+                            .cast<StringAttr>()
+                            .getValue()
+                     << "\n";
+    }
+  });
 
-    auto &info = interfaceMap[{defName, name}];
-    if (info.isMissing()) {
-      emitCircuitError()
-          << "'firrtl.circuit' op contained a Grand Central Interface '"
-          << defName << "' that had an element '" << name
-          << "' which did not have a scattered companion annotation (is there "
-             "an invalid target in your annotation file?)";
-      continue;
+  // Make an InstancePaths object available to methods of the Grand Central
+  // pass.
+  //
+  // TODO: There should be a better way to do this.
```

Comment:
Move semantics would let you construct an empty one, the construct a new one over the top.

Reply:
Or make it an optional to allow late construction in place.

Reply:
I like the optional approach to lazily construct this. However, there are problems with that as `BumpPtrAllocator` has a deleted copy constructor.  I'll use the lazy construction approach for the `CircuitNamespace` and leave instance paths lazy construction as future work.